src/expr/src/relation/canonicalize.rs

```diff
@@ -83,6 +87,123 @@ pub fn canonicalize_predicates(predicates: &mut Vec<MirScalarExpr>, input_type:
+    // `true` (resp. `false`) every subexpression in
+    // `(p1 & ... & p(i-1) & p(i+1) ... & pn)` that is equal to `pi`.
+
+    // (As for the reduction does not work if `p` can be null: if `p` is null,
+    // then p & f(p) becomes `null & f(null)`. `null & f(null)` is not
+    // necessarily equal to `null & f(true)` because `null & true = null`
+    // and `null & false = false`.)
+
+    // Note that this does some dedupping of predicates since if `p1 = p2`
+    // then this reduction process will replace `p1` with true.
+
+    // Pop out a predicate `p` from `predicates_to_apply`. Check all the other
+    // predicates (this is the union of `predicates_to_apply` and
+    // `applied_predicates`), and if `p` is a subexpression of any of those
+    // predicates, replace the subexpression with `true`. Then stick `p` into
+    // `applied_predicates`. If a predicate `q` in `applied_predicates` was
+    // altered in the aforementioned process, remove `q` from
+    // `applied_predicates` and put it in `predicates_to_apply` so we can reduce
+    // the predicates even further.
+    let mut predicates_to_apply = Vec::new();
+    let mut applied_predicates = Vec::<MirScalarExpr>::new();
+    // Seed `predicates_to_apply` with the `Vec` of all predicates.
+    std::mem::swap(&mut predicates_to_apply, predicates);
+
+    while let Some(predicate_to_apply) = predicates_to_apply.pop() {
+        let mut replace_subexpr_other_predicates =
+            |expr: &MirScalarExpr, constant_bool: &MirScalarExpr| {
+                // Do not replace subexpressions equal to `expr` if `expr` is a
+                // literal to avoid infinite looping.
+                if !expr.is_literal() && !expr.typ(input_type).nullable {
+                    let replace_subexpr_other_predicate = |other_predicate: &mut MirScalarExpr| {
+                        let mut changed = false;
+                        visit_mut_replace_subexpr(
+                            other_predicate,
+                            &mut |e| replace(e, expr, constant_bool),
+                            &mut changed,
+                        );
+                        if changed {
+                            other_predicate.reduce(input_type);
+                        }
+                        changed
+                    };
+                    for other_predicate in predicates_to_apply.iter_mut() {
+                        replace_subexpr_other_predicate(other_predicate);
+                    }
+                    for other_idx in (0..applied_predicates.len()).rev() {
+                        if replace_subexpr_other_predicate(&mut applied_predicates[other_idx]) {
+                            predicates_to_apply.push(applied_predicates.remove(other_idx));
+                        }
+                    }
+                }
+            };
+
+        if let MirScalarExpr::CallUnary {
+            func: UnaryFunc::Not,
+            expr,
+        } = &predicate_to_apply
```

Comment:
For my style preferences, this would be a great place for a comment to call out that this is the meat of the method. Somehow having read through everything above, I didn't realize that this is where the behavior for this method actually starts.